Filtered_kernel/include/CGAL/Lazy.h

```diff
@@ -231,6 +229,11 @@ class Lazy_rep : public Rep, public Depth_base
 
   mutable AT at;
   mutable ET *et;
+#ifdef CGAL_LAZY_USE_MUTEX
+  typedef CGAL::cpp11::mutex Mutex;
+  typedef CGAL::cpp11::lock_guard<Mutex> Lock_guard;
+  mutable Mutex update_exact_mutex;
+#endif
```

Comment:
Just thinking out loud, I haven't looked at Lazy in a couple years. Wouldn't it be possible to implement this lock-free? We could compute the exact value, and atomically set a pointer to it once it is ready (in pathological cases, we might end up computing this value several times, but I doubt it will ever be significant). This is not as easy as it sounds. If we work at the level of the Lazy object (includes the approximate value and a pointer to the exact one), there are lifetime issues (a thread trying to read the approximate value could read the old pointer and have another thread destroy the old object before the first one reads the approximate value). Doing it at the level of the exact object seems safer (the pointer can only go from 0 to non-zero if we use CAS), but then we probably also need to read/write the approximate value atomically (theoretically at least).

By the way, did you look at more fine-grained locking? Shared locking for readers and unique locking for writers comes to mind.